File: 2023/UICPC DIV1/UICPC Round #03/solutions/H/pegsandlegs-zf.py
#!/usr/bin/python3

L, P = map(int, input().split())

hit = set()
dp = {}

for i in range(L+P):
    l = input().split()
    if i < L:
        dp[i] = (float(l[0]), 0.0)
    else:
        pl, pr = float(l[0]), float(l[1])
        il, ir = int(l[2])-1, int(l[3])-1
        dp[i] = (pl*dp[il][0] + pr*dp[ir][0], 1.0-pl-pr + pl*dp[il][1] + pr*dp[ir][1])
        hit.add(il)
        hit.add(ir)

best = 0.0
for i in dp:
    if i not in hit:
        best = max(best, dp[i][0]/(1.0-dp[i][1]))

print("{0:.10f}".format(best))

File: 2023/UICPC DIV1/UICPC Round #03/solutions/H/pegsandlegs-zf.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
  int L, P;
  cin >> L >> P;
  int n = L+P;

  vector<double> e(n), p(n, 0.0);
  vector<bool> hit(n, false);
  for (int i = 0; i < L+P; ++i) {
    if (i < L) cin >> e[i];
    else {
      double pl, pr;
      int il, ir;
      cin >> pl >> pr >> il >> ir;
      --il; --ir;
      e[i] = pl*e[il] + pr*e[ir];
      p[i] = 1.0-pl-pr + pl*p[il] + pr*p[ir];
      hit[il] = hit[ir] = true;
    }
  }

  double best = 0.0;
  for (int i = 0; i < n; ++i) if (!hit[i]) best = max(best, e[i]/(1-p[i]));

  cout << fixed << setprecision(10) << best << endl;
}
